src/types.py

```diff
@@ -239,67 +325,172 @@ def get_function_pointer_signature(self) -> Optional["FunctionSignature"]:
+            # Check if there is a field at the `remaining_offset` offset
+            subpath, subtype, sub_remainder_bits = data.ptr_to.get_field(
+                remaining_offset, target_size=target_size
+            )
+            if subpath is not None:
+                # Success: prepend `index` and return
+                subpath.insert(0, index)
+                return subpath, subtype, sub_remainder_bits
+            return NO_MATCHING_FIELD
+
+        if self.is_struct():
+            data = self.data()
+            assert data.struct is not None
+            possible_fields = data.struct.fields_at_offset(offset)
+            if not possible_fields:
+                return NO_MATCHING_FIELD
+            possible_results: List[Type.GetFieldResult] = []
+            if target_size is None or target_size == self.get_size_bytes():
+                possible_results.append(([], self, offset))
+            for field in possible_fields:
+                inner_offset_bits = offset - field.offset
+                subpath, subtype, sub_remainder_bits = field.type.get_field(
+                    inner_offset_bits, target_size=target_size
+                )
+                if subpath is None:
+                    continue
+                subpath.insert(0, field.name)
+                possible_results.append((subpath, subtype, sub_remainder_bits))
+                if (
+                    target_size is not None
+                    and target_size == subtype.get_size_bytes()
+                    and sub_remainder_bits == 0
+                ):
+                    return possible_results[-1]
+            zero_offset_results = [r for r in possible_results if r[2] == 0]
+            if zero_offset_results:
+                return zero_offset_results[0]
+            if possible_results:
+                return possible_results[0]
+
+        if offset == 0 and (
+            target_size is None or target_size == self.get_size_bytes()
+        ):
+            # The whole type itself is a match
+            return [], self, 0
+
+        return NO_MATCHING_FIELD
+
+    def get_deref_field(
+        self, offset: int, *, target_size: Optional[int]
+    ) -> GetFieldResult:
+        """
+        Similar to `.get_field()`, but treat self as a pointer and find the field in the
+        pointer's target.  The return value has the same semantics as `.get_field()`.
+
+        If successful, the first item in the resulting `field_path` will be `0`.
+        This mirrors how `foo[0].bar` and `foo->bar` are equivalent in C.
+        """
+        NO_MATCHING_FIELD: Type.GetFieldResult = (None, Type.any(), offset)
+
+        target = self.get_pointer_target()
+        if target is None:
+            return NO_MATCHING_FIELD
+
+        # Assume the pointer is to a single object, and not an array.
+        size = target.get_size_bytes()
+        if offset < 0 or (size is not None and offset >= size):
```

Comment:
offset < 0 will be checked later by get_field fwiw, and I suspect offset >= size implicitly also is. So I think this check may be redundant